[Merged by Bors] - refactor: allow for bilinear maps of norm greater than one in Hölder's inequality and use `HolderTriple`

j-loreaux opened the pull request:
This PR does two things, both related to statements of Hölder's inequality:

1. In the generic statements involving a general map `b : E → F → G` (which we think of as a continuous bilinear map despite the fact that these are not quite the hypotheses), we loosen the assumptions on `b` so that it need not have norm bounded by 1. This makes it possible to use general continuous bilinear maps in the forthcoming #21583, and addresses the concerns in #6108.
2. Changes the statements to make use of the new `ENNReal.HolderTriple` type class. Among other things, this makes some existing lemmas specialized to `∞` redundant, so we deprecate them.

Closes #6108.

---

- [x] depends on: #21854

If this doesn't completely address #6108, let me know and I can do the rest of what you wanted.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

YaelDillies commented on Mathlib/MeasureTheory/Function/LpSeminorm/CompareExp.lean:
```diff
@@ -286,28 +273,28 @@ variable {𝕜 α E F : Type*} {m : MeasurableSpace α} {μ : Measure α} [Norme
-    (Eventually.of_forall fun _ => nnnorm_smul_le _ _) :)
+    (φ : α → 𝕜) : eLpNorm (φ • f) p μ ≤ eLpNorm φ ∞ μ * eLpNorm f p μ := by
+  simpa using (eLpNorm_le_eLpNorm_top_mul_eLpNorm p φ hf (· • ·) 1
+    (Eventually.of_forall fun _ => by simpa using nnnorm_smul_le _ _) :)
```
```suggestion
    (.of_forall fun _ => by simpa using nnnorm_smul_le _ _) :)
```
if you want while you're at it

YaelDillies commented:
Is there any chance you can split this PR into:
* "allow for bilinear maps of norm greater than one in Hölder's inequality"
* "use `HolderTriple`"

? The current diff is quite hard to read and the two changes seem essentially disjoint

j-loreaux commented:
I don't think the diff being hard to read would change much if I split it. The "use Hölder triple" changes are pretty minor in terms of diff size (but I would essentially have to recreate them from scratch if I were to try and split the PR).

The main issue for the diff is that when generalizing to bilinear maps, I found it much easier to reprove some of the results from scratch (and in at least one case, shorten the proof considerably), by taking another approach. For example, I used `ENNReal.trichotomy` in one proof to give myself nicer case splits, and I stayed in the realm of `eLpNorm` before passing to the integral in another in order to make it easier to pull the constant out.

YaelDillies commented on Mathlib/MeasureTheory/Function/LpSeminorm/CompareExp.lean:
Do you think we should keep the `c = 1` lemmas as convenient special cases?

j-loreaux replied:
no, as you can see with the `mul` and `smul` results, it seems to only add an extra invocation of `simp`, so it's minor, and I think those are currently the only places the results are used in Mathlib.

YaelDillies replied:
Would you mind commenting that on https://github.com/fpvandoorn/carleson/pull/231? @js2357 is similarly adding `c = 1` special cases there

j-loreaux replied:
oops, sorry, I missed this.

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

RemyDegenne commented:
bors r+
